### ISSUE TITLE
Subset scattering, decimation

### ISSUE THREAD

**Issue by OverLordGoldDragon** [kymatio/kymatio#791]
1. Sometimes we know where most energy will be and discard coefficients as a post-processing step. If they happen to be high-freq coeffs, however, they eat up majority of compute - and it'd be much more effective to not compute them at all. This shouldn't be hard to implement - worst part will be adjusting meta & tests.

 2. Decimating grants speed & memory without changing the scale of invariance. Often, the final output barely differs from undecimated. For this we can introduce `Td` (`T` decimate) and `Fd`. A problem is introducing negatives at output - but this can be easily compensated for with an offset.

I use both these with success. Here's a code excerpt for 2:

```python
sub_T = 2**(log2_T - k2)
o = fft(o, axis=-1) * phi_t[k2]
o = o.reshape(len(o), sub_T, o.shape[-1]//sub_T).mean(axis=1)
o = ifft(o, axis=-1).real
if Td > 1:
    o = scipy.signal.decimate(o, Td, ftype='fir', axis=-1)
```
This is huge for JTFS.

**Comment by OverLordGoldDragon** [comment]
Spinned JTFS coefficient example:

<img src="https://user-images.githubusercontent.com/16495490/150319698-dba51f80-d7c8-4507-93da-c490841144ee.gif" width="200">

**Comment by lostanlen** [comment]
Yes. But decimated relative to _what_?

There's nothing in this that we couldn't do with an appropriate API for `stride` and `bins_per_octave`
`stride = Td *T/2` and `bins_per_octave = Q/Fd`

TL;DR this is important but we should be careful to express it in a unit that is intuitive. I don't want 2 more kwargs that default to `None`

**Comment by OverLordGoldDragon** [comment]
> Yes. But decimated relative to what?

Decimated right after lowpassing. Else, I don't get the question

> bins_per_octave = Q/Fd

This isn't same: `compute(downsampled) != downsample(computed)`.

> stride = Td *T/2

Dunno what the 2's about but 1) changing `T` changes scale of invariance; 2) simply increasing stride is akin to excess subsampling, which isn't equivalent to decimating

> I don't want 2 more kwargs that default to None

They can't be done as post-processing steps, and can improve speed & memory at least two-fold. If we're getting overwhelmed by kwargs, we could define an optional `cfg` dictionary arg that stores all the "advanced" configs - less namespace clutter, more unlimited features.

**Comment by lostanlen** [comment]
But then you change the scale of invariance, since `scipy.signal.decimate` applies an anti-aliasing filter, i.e. a low-pass filter.

**Comment by OverLordGoldDragon** [comment]
Yes, but it's a different low-pass, with much less blur, as shown by the GIF. It more faithfully captures the original.

**Comment by lostanlen** [comment]
I see. Then it's more a problem of `phi` design.
Right now we have `phi`= Gaussian, which makes energy conservation ~easy~ _easier_ but may not be the only reasonable choice.
By default, `decimate` uses an order 8 Chebyshev type I filter instead. That's only one of many possible choices.

I guess it's a tradeoff between aliasing and blurring?

If the profile of `phi` truly makes a different in classification (there is some evidence it does, see LEAF by Zeghidour et al.) then i imagine that the best option is to set `T` to a low value and then learn a non-Gaussian average pooling operator on top.

**Comment by OverLordGoldDragon** [comment]
> Then it's more a problem of phi design.

No. The primary role of the Gaussian is to impose invariance; downsampling is a bonus. The decimating lowpass isn't suited for invariance. One _can_ change `T` purely for sake of downsampling, but shouldn't: the two steps should be separate.

> then i imagine that the best option is to set T to a low value and then learn

That's a choice the user can always make. This doesn't solve the case when we need a specific `T` but can't afford it per speed or memory, however.

To emphasize, the purpose is to enable superior downsampling _purely for sake of downsampling_, while minimally affecting the scale of invariance. This is especially important in JTFS, as we often really don't want a higher `F` invariance, but do want a higher `F` downsampling, and can afford generous amounts of it with minimal loss of information.

**Comment by lostanlen** [comment]
OK, then i'm completely lost. What are we trying to do?

**Comment by OverLordGoldDragon** [comment]
> Sounds like we should just downsample more by default then?

Maybe, but I'd best leave it to the user

> Are you saying you want stride > T ?

Exactly

**Comment by MuawizChaudhary** [comment]
Scattering iterators #735 will solve everything inshahallah
